src/lib/kadm5/clnt/client_init.c

```diff
@@ -221,9 +221,16 @@ init_any(krb5_context context, char *client_name, enum init_type init_type,
         return KADM5_MISSING_KRB5_CONF_PARAMS;
     }
 
+    /*
+     * Parse the client name.  If it has a realm, it is almost certainly a
+     * host-based principal using DNS fallback processing or the referral
+     * realm, so give it the appropriate name type for canonicalization.
+     */
     code = krb5_parse_name(handle->context, client_name, &client);
     if (code)
         goto error;
+    if (client->realm.length == 0)
+        client->type = KRB5_NT_SRV_HST;
```

Comment:
Does this belong in `k5_infer_principal_type()`?

Reply:
An interesting question.  I think the answer is no: in libkadm5clnt we know a bit about the inputs (although not with super-high confidence, since nothing stops someone from running "kadmin -k -p ghudson@"), but more importantly, we know the name is for an AS client principal, so it can't possibly work with an empty realm.  So it's safe to assume that it's a host-based principal from sn2princ, because if it's not it wasn't going to work anyway.  Putting the logic in k5_infer_principal_type() would be much less bounded.

Put another way: this is a hack because we don't want to redesign the libkadm5 initialization APIs, and we can show that it's a safe hack, but it doesn't belong at so low a level as the library functions for building or parsing a principal name.

Reply:
A related question is whether we should also conditionalize on `init_type == INIT_SKEY`, to better scope the hack.  I'm leaning towards yes, but practically it shouldn't matter much either way.